core/crypto/signed_hdr.c

```diff
@@ -48,21 +48,31 @@ struct shdr *shdr_alloc_and_copy(size_t offs, const void *img, size_t img_size)
 	return shdr;
 }
 
+static bool is_strong_hash_algo(uint32_t algo)
+{
+	return algo == TEE_ALG_SHA224 || algo == TEE_ALG_SHA256 ||
+	       algo == TEE_ALG_SHA384 || algo == TEE_ALG_SHA512;
+}
```

Comment:
How about excluding the known weak algorithms instead? There would be only two entries here, and the code would only need changing when an algorithm becomes weak, not when a new (strong) algorithm is added.

Reply:
Assume that we check against weak algorithms instead. Should we check against MD5 even if it can occur?

It seems more clear to check which algorithms we're prepared to accept.

Reply:
> Assume that we check against weak algorithms instead. Should we check against MD5 even if it can occur?

Yes.

> It seems more clear to check which algorithms we're prepared to accept.

...which is "*all* algorithms except the weak ones". For example, GP 1.3.1 defines `TEE_ALG_SHA3_{224,256,384,512}` which would be perfectly acceptable if OP-TEE implemented them. So with the following:

```C
static bool is_weak_hash_algo(uint32_t algo)
{
	return algo == TEE_ALG_MD5 || algo == TEE_ALG_SHA1;
}
```
...we could add support for SHA3 without changing anything here.

Reply:
I'm sorry, I didn't see your reply until now.
I suppose it should be safe to assume that we're not going to add any new weak algorithms, only strong ones.

Reply:
We have a definition for `TEE_ALG_MD5SHA1` too. I'm not sure what to make of that though.

Reply:
> I suppose it should be safe to assume that we're not going to add any new weak algorithms, only strong ones.

Yes, and that is the reason for my proposal. One of us must be missing something obvious in this discussion :-/
Let me try to state it again: if we're not going to add any new weak algorithm (we agree on that), why impose to change this code every time a new strong algorithm is added (your proposal), when we could just do nothing (mine)?

Reply:
> We have a definition for `TEE_ALG_MD5SHA1` too. I'm not sure what to make of that though.

Ha! Indeed. Non-standard algo. It is MD5 and SHA-1 concatenated, so if both are considered weak I guess it should be considered weak too. The hashes are not even combined in any subtle way.